pkg/streamingpromql/operators/scalar_constant.go

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type ScalarConstant struct {
+	Value                    float64
+	Start                    int64 // Milliseconds since Unix epoch
+	End                      int64 // Milliseconds since Unix epoch
+	Interval                 int64 // In milliseconds
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	expressionPosition posrange.PositionRange
+}
+
+var _ types.ScalarOperator = &ScalarConstant{}
+
+func NewScalarConstant(
+	value float64,
+	start int64,
+	end int64,
+	interval int64,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+) *ScalarConstant {
+	return &ScalarConstant{
+		Value:                    value,
+		Start:                    start,
+		End:                      end,
+		Interval:                 interval,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+func (s *ScalarConstant) GetValues(_ context.Context) (types.ScalarData, error) {
+	numSteps := stepCount(s.Start, s.End, s.Interval)
+	samples, err := types.FPointSlicePool.Get(numSteps, s.MemoryConsumptionTracker)
+
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	samples = samples[:numSteps]
+
+	for step := 0; step < numSteps; step++ {
+		samples[step].T = s.Start + int64(step)*s.Interval
+		samples[step].F = s.Value
+	}
+
+	return types.ScalarData{Samples: samples}, nil
```

Comment:
I wonder (and this probably applies for other operators too) if we run the risk of forgetting to put something back into the pool. Would it be better for each operator to use its `Close` to return items to the pool once it is done with them? We could have GetValues return a closer that the caller needs to handle if the timing is difficult.

Reply:
>if we run the risk of forgetting to put something back into the pool

It's not ideal if we forget to put something back into the pool, but the impact is that we'll just allocate more slices more often than we'd otherwise need to do.

>Would it be better for each operator to use its `Close` to return items to the pool once it is done with them? 

The downside of this is that an operator doesn't know how its slices will be used, and each operator would need to then track all slices it returns, which would keep them alive and increase peak memory consumption.

For example, let's say operator A returns a slice of `FPoint`s to operator B.

If operator B is an operation like `-metric`, it could mutate that slice in place and pass it to the operator consuming it, without getting another slice. If both operators implemented the "return to pool on `Close`" pattern, then both A and B would return the same slice to the pool, so it would be used twice (pools don't do any kind of deduplication). 

Furthermore, imagine operator A returns many series - it would then need to hold a reference to the slice for each series so that it can return them to the pool when `Close` is called.

Lastly, delaying returning slices to the pool until `Close` delays returning those slices until the query is complete. This means those slices can't be reused by the same query, but a single query tends to need many slices of the same size, one for each series, so this would be a great candidate for pooling.

Reply:
Thanks for the detailed examples. I'm aware there would be extra overhead, but I'm wondering if we can add some more safeguards in place. Unfortunately I don't have any concrete ideas how this might look. Perhaps some kind of test that can check everything has been put back at the end, or a way of asserting that if you receive a value from this (or similar methods) the caller is responsible for putting xyz back. Anyway, this is probably a discussion outside of this PR, but it's something to keep in mind.

Reply:
>Perhaps some kind of test that can check everything has been put back at the end, 

I've been thinking about this as well - perhaps we could add a test to ensure that the estimated current memory consumption is 0 after a query is complete. If it was negative, then we've returned something to the pool multiple times, and if it was positive, then we've missed returning something to the pool.

I've also wondered if we should just panic if the estimated current memory consumption goes negative, as that would indicate something being returned multiple times and would almost certainly lead to incorrect query results.

But yep, I think this is something for another PR.

>or a way of asserting that if you receive a value from this (or similar methods) the caller is responsible for putting xyz back

The convention we've been following is that if you receive a slice from an operator (samples or metadata), you can mutate it as you please, and you're responsible for returning the slice to the pool when you're done (or passing it to something else that has the same responsibility).

Reply:
> But yep, I think this is something for another PR.

Sounds good :-).

> The convention we've been following is that if you receive a slice from an operator (samples or metadata), you can mutate it as you please, and you're responsible for returning the slice to the pool when you're done (or passing it to something else that has the same responsibility).

Yep, but I meant a way of enforcing or checking that convention.